dnscrypt.c

```diff
@@ -206,33 +217,46 @@ dnscrypt_server_uncurve(struct context *c, const KeyPair *keypair,
     struct dnscrypt_query_header *query_header =
         (struct dnscrypt_query_header *)buf;
     memcpy(nmkey, query_header->publickey, crypto_box_PUBLICKEYBYTES);
-    if (crypto_box_beforenm(nmkey, nmkey, keypair->crypt_secretkey) != 0) {
-        return -1;
+    if (use_xchacha20) {
+#ifdef HAVE_CRYPTO_BOX_CURVE25519XCHACHA20POLY1305_OPEN_EASY
+        if (crypto_box_curve25519xchacha20poly1305_beforenm(nmkey, nmkey, keypair->crypt_secretkey) != 0) {
+            return -1;
+        }
+#endif
+    } else {
+        if (crypto_box_beforenm(nmkey, nmkey, keypair->crypt_secretkey) != 0) {
+            return -1;
+        }
     }
 
     uint8_t nonce[crypto_box_NONCEBYTES];
     memcpy(nonce, query_header->nonce, crypto_box_HALF_NONCEBYTES);
     memset(nonce + crypto_box_HALF_NONCEBYTES, 0, crypto_box_HALF_NONCEBYTES);
 
-    memset(buf + DNSCRYPT_QUERY_BOX_OFFSET - crypto_box_BOXZEROBYTES, 0,
-           crypto_box_BOXZEROBYTES);
-    if (crypto_box_open_afternm
-        (buf + DNSCRYPT_QUERY_BOX_OFFSET - crypto_box_BOXZEROBYTES,
-         buf + DNSCRYPT_QUERY_BOX_OFFSET - crypto_box_BOXZEROBYTES,
-         len - DNSCRYPT_QUERY_BOX_OFFSET + crypto_box_BOXZEROBYTES, nonce,
-         nmkey) != 0) {
-        return -1;
+    if (use_xchacha20) {
+#ifdef HAVE_CRYPTO_BOX_CURVE25519XCHACHA20POLY1305_OPEN_EASY
+        if (crypto_box_curve25519xchacha20poly1305_open_easy_afternm
+            (buf, buf + DNSCRYPT_QUERY_BOX_OFFSET,
+             len - DNSCRYPT_QUERY_BOX_OFFSET, nonce, nmkey) != 0) {
+            return -1;
+        }
+#endif
+    } else {
+        if (crypto_box_open_easy_afternm
```

Comment:
apparently this is what made the original build to fail (along with sodium_increment v1.0.4). I would be handy to specify the required version for libsodium in the README. This way it makes it easier for people to see if they need to compile from source or if they can use the distro version.
CentOS 7 currently ships with 1.0.5 for instance, which means that dnscrypt-wrapper can be compile using the distro version.